### PR TITLE
Added vnodes index partitioning.

### PR DESCRIPTION
**Pull request by ealonsodb**

### CODE REVIEW

File: builder/src/main/java/com/stratio/cassandra/lucene/builder/index/Partitioner.java [hunk 1]
```diff
@@ -104,4 +106,31 @@ public OnColumn(int partitions, String column) {
...
+     * The number of partitions per node should be specified.
+     */
+    public static class OnVirtualNodes extends Partitioner {
+
+        /** The number of partitions per node. */
+        @JsonProperty("partitions")
+        public final int partitions;
+
+        /**
+         * Builds a new partitioner on virtual nodes token ranges.
```
Comment (ealonsodb) [inline builder/src/main/java/com/stratio/cassandra/lucene/builder/index/Partitioner.java]: change comment to same

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,7 @@
 ## 3.0.10.3 (Upcoming)
 
 * Fix deletion of ranges of rows in a partition
+* Add virtual nodes token ranges based index partitioner
```
Comment (ealonsodb) [inline CHANGELOG.md]: virtual node based pàrttioner

File: builder/src/main/java/com/stratio/cassandra/lucene/builder/index/Partitioner.java [hunk 3]
```diff
@@ -104,4 +106,31 @@ public OnColumn(int partitions, String column) {
             this.column = column;
         }
     }
+
+    /**
+     * A {@link Partitioner} based on the partition key token. Rows will be stored in an index partition determined by
+     * the virtual nodes token range. Partition-directed searches will be routed to a single partition, increasing
```
Comment (ealonsodb) [inline builder/src/main/java/com/stratio/cassandra/lucene/builder/index/Partitioner.java]: by the virtul node.

File: builder/src/main/java/com/stratio/cassandra/lucene/builder/index/Partitioner.java [hunk 4]
```diff
@@ -104,4 +106,31 @@ public OnColumn(int partitions, String column) {
...
+     * performance. However,unfiltered token range searches will be routed to all the partitions, with a slightly lower
```
Comment (ealonsodb) [inline builder/src/main/java/com/stratio/cassandra/lucene/builder/index/Partitioner.java]: unfiltered -> unbouded

File: builder/src/main/java/com/stratio/cassandra/lucene/builder/index/Partitioner.java [hunk 5]
```diff
@@ -104,4 +106,31 @@ public OnColumn(int partitions, String column) {
...
+     * performance. Virtual node token range queries will be routed to only one partition which increase performance in
+     * spark queries with vnodes rather than partitioning on token.
+     *
+     * This partitioner load balance depends on virtual node token ranges asignation. The more virtual nodes, the better
+     * distribution (more similarity in number of tokens that falls inside any virtual node) between virtual nodes, the
+     * better load balnce with this partitioner.
```
Comment (ealonsodb) [inline builder/src/main/java/com/stratio/cassandra/lucene/builder/index/Partitioner.java]: balance

File: doc/documentation.rst [hunk 6]
```diff
@@ -15,6 +15,7 @@ Stratio's Cassandra Lucene Index
         - `None partitioner <#none-partitioner>`__
         - `Token partitioner <#token-partitioner>`__
         - `Column partitioner <#column-partitioner>`__
+        - `Virtual nodes partitioner <#virtual-nodes-partitioner>`__
```
Comment (ealonsodb) [inline doc/documentation.rst]: nodes -> node

File: doc/documentation.rst [hunk 7]
```diff
@@ -709,6 +710,49 @@ cardinalities and uniform distributions will provide better load balancing betwe
 
     SELECT * FROM tweets WHERE expr(idx, '{...}')'; -- Fetches all nodes, all partitions
 
+Virtual nodes partitioner
```
Comment (ealonsodb) [inline doc/documentation.rst]: nodes-> node

File: doc/documentation.rst [hunk 8]
```diff
@@ -709,6 +710,49 @@ cardinalities and uniform distributions will provide better load balancing betwe
...
+_________________________
+
+A partitioner based on the virtual nodes partition key token ranges. Rows will be stored in an index partition determined
```
Comment (ealonsodb) [inline doc/documentation.rst]: nodes-> node

File: doc/documentation.rst [hunk 9]
```diff
@@ -709,6 +710,49 @@ cardinalities and uniform distributions will provide better load balancing betwe
...
+       id INT,
+       body TEXT
+       PRIMARY KEY ((user, month), date, id)
+    );
+
+    CREATE CUSTOM INDEX idx ON tweets()
+    USING 'com.stratio.cassandra.lucene.Index'
+    WITH OPTIONS = {
+       'schema': '{...}',
+       'partitioner': '{type: "vnodes", partitions: 4}',
```
Comment (ealonsodb) [inline doc/documentation.rst]: vnodes-> vnode

File: plugin/src/main/scala/com/stratio/cassandra/lucene/IndexService.scala [hunk 10]
```diff
@@ -283,6 +283,7 @@ abstract class IndexService(
     // Search
     tracer.trace(s"Lucene index searching for $count rows")
     val partitions = partitioner.partitions(command)
+    logger.debug(s"searching in partitions ${partitions.foldLeft("")((b,a) => b + "" + a.toString)}")
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/IndexService.scala]: remove

File: plugin/src/main/scala/com/stratio/cassandra/lucene/index/PartitionedIndex.scala [hunk 11]
```diff
@@ -158,7 +158,7 @@ class PartitionedIndex(
     * @param document  the document to be added
     */
   def upsert(partition: Int, term: Term, document: Document) {
-    logger.debug(s"Indexing $document with term $term in $name")
+    logger.debug(s"Indexing $document with term $term in $name in partition number $partition")
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/index/PartitionedIndex.scala]: rempove debugs

File: plugin/src/main/scala/com/stratio/cassandra/lucene/index/PartitionedIndex.scala [hunk 12]
```diff
@@ -178,7 +178,7 @@ class PartitionedIndex(
     * @param query     the query identifying the documents to be deleted
     */
   def delete(partition: Int, query: Query) {
-    logger.debug(s"Deleting $query from $name")
+    logger.debug(s"Deleting $query from $name in parttion $partition")
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/index/PartitionedIndex.scala]: parttion -> aprtttiition

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/Partitioner.scala [hunk 13]
```diff
@@ -79,7 +79,8 @@ object Partitioner {
   @JsonSubTypes(Array(
     new JsonSubTypes.Type(value = classOf[PartitionerOnNone.Builder], name = "none"),
     new JsonSubTypes.Type(value = classOf[PartitionerOnToken.Builder], name = "token"),
-    new JsonSubTypes.Type(value = classOf[PartitionerOnColumn.Builder], name = "column")))
+    new JsonSubTypes.Type(value = classOf[PartitionerOnColumn.Builder], name = "column"),
+    new JsonSubTypes.Type(value = classOf[PartitionerOnVirtualNodes.Builder], name = "vnodes")))
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/Partitioner.scala]: vnodes-> vnode

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+import com.stratio.cassandra.lucene.IndexException
+import com.stratio.cassandra.lucene.util.Logging
+import org.apache.cassandra.config.CFMetaData
+import org.apache.cassandra.db._
+import org.apache.cassandra.dht.Token
+import org.apache.cassandra.service.StorageService
+
+import scala.collection.JavaConverters._
+
+/** [[Partitioner]]  based on the partition key token. Rows will be stored in an index partition
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala]: remove double space

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+  * determined by the virtual nodes token range. Partition-directed searches will be routed to a
+  * single partition, increasing performance. However,unfiltered token range searches will be routed
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala]: ,uniflter -> , unfiltered

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+  * than partitioning on token.
+  *
+  * This partitioner load balance depends on virtual node token ranges asignation. The more virtual
+  * nodes, the better distribution (more similarity in number of tokens that falls inside any virtual
+  * node) between virtual nodes, the better load balance with this partitioner.
+  *
+  * @param partitions_number the number of index partitions per node
+  * @author Eduardo Alonso `eduardoalonso@stratio.com`
+  */
+case class PartitionerOnVirtualNodes(partitions_number: Int) extends Partitioner with Logging {
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala]: patttions_number  es una puta mierda y lo odio

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    val upperPartition = partition(upper)
+
+    if (lowerPartition <= upperPartition)
+      (lowerPartition to upperPartition).toList
+    else
+      (lowerPartition to numTokens).toList ::: (0 to upperPartition).toList
+  }
+
+
+  /** @inheritdoc*/
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala]: space needed

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+  override def partition(key: DecoratedKey): Int = partition(key.getToken)
+
+  /** @inheritdoc */
+  override def partitions(command: ReadCommand): List[Int] = command match {
+    case c: SinglePartitionReadCommand => List(partition(c.partitionKey))
+    case c: PartitionRangeReadCommand =>
+      val range = c.dataRange()
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala]: remove parenthesiss

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+      val start = range.startKey().getToken
+      val stop = range.stopKey().getToken
+
+      if (start.equals(stop)) {
+        if (start.isMinimum) {
+          //this is a unbounded select
+          allPartitions
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala]: include this logic in partttions(start, stop)

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+    *
+    * @param token the token Long value
+    * @return the virtual node index where this token falls into
+    */
+  private[this] def virtualNode(token: Long): Int = {
+    val vnode = tokens.count(x => token >= x) - 1
+    if (vnode < 0) numTokens + vnode else vnode
+  }
+
+  /** @inheritdoc*/
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNodes.scala]: space

File: plugin/src/test/scala/com/stratio/cassandra/lucene/IndexOptionsTest.scala [hunk 21]
```diff
@@ -16,7 +16,7 @@
 package com.stratio.cassandra.lucene
 
 import com.stratio.cassandra.lucene.IndexOptions._
-import com.stratio.cassandra.lucene.partitioning.{PartitionerOnNone, PartitionerOnToken}
+import com.stratio.cassandra.lucene.partitioning.{PartitionerOnNone, PartitionerOnToken, PartitionerOnVirtualNodes$}
```
Comment (ealonsodb) [inline plugin/src/test/scala/com/stratio/cassandra/lucene/IndexOptionsTest.scala]: remove bad import

File: plugin/src/main/scala/com/stratio/cassandra/lucene/index/PartitionedIndex.scala [hunk 22]
```diff
@@ -178,7 +178,7 @@ class PartitionedIndex(
...
     indexes(partition).delete(query)
   }
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/index/PartitionedIndex.scala]: line 197 chunk cannot be parttitions:?????

File: doc/documentation.rst [hunk 23]
```diff
@@ -709,6 +710,47 @@ cardinalities and uniform distributions will provide better load balancing betwe
...
+       id INT,
+       body TEXT
+       PRIMARY KEY ((user, month), date, id)
+    );
+
+    CREATE CUSTOM INDEX idx ON tweets()
+    USING 'com.stratio.cassandra.lucene.Index'
+    WITH OPTIONS = {
+       'schema': '{...}',
+       'partitioner': '{type: "vnode", partitions: 4}',
```
Comment (ealonsodb) [inline doc/documentation.rst]: fix this

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNode.scala [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+  * nodes, the better distribution (more similarity in number of tokens that falls inside any virtual
+  * node) between virtual nodes, the better load balance with this partitioner.
+  *
+  * @param vnodes_per_partition the number of virtual nodes that falls inside an index partition
+  * @author Eduardo Alonso `eduardoalonso@stratio.com`
+  */
+case class PartitionerOnVirtualNode(vnodes_per_partition: Int) extends Partitioner with Logging {
+
+  if (vnodes_per_partition <= 0) throw new IndexException(
+    s"The number of partitions should be strictly positive but found $vnodes_per_partition")
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNode.scala]: fix number of parttitions-> vnodes_per_parttition

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNode.scala [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+
+  val tokens: List[Token] = StorageService.instance.getLocalTokens.asScala.toList
+  val numTokens = tokens.size
+
+  /** @inheritdoc */
+  override def numPartitions : Int = (numTokens.toDouble/vnodes_per_partition.toDouble).ceil.toInt
+
+  if (numTokens == 1) logger.warn("You are using a PartitionerOnVirtualNode but cassandra is only configured with one token (non using virtual nodes.)")
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNode.scala]: non-> nopt .) -> ).

File: plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNode.scala [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+  *
+  * @param vnodes_per_partition the number of virtual nodes that falls inside an index partition
+  * @author Eduardo Alonso `eduardoalonso@stratio.com`
+  */
+case class PartitionerOnVirtualNode(vnodes_per_partition: Int) extends Partitioner with Logging {
+
+  if (vnodes_per_partition <= 0) throw new IndexException(
+    s"The number of partitions should be strictly positive but found $vnodes_per_partition")
+
+  val tokens: List[Token] = StorageService.instance.getLocalTokens.asScala.toList
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/partitioning/PartitionerOnVirtualNode.scala]: sort lista

File: plugin/src/main/scala/com/stratio/cassandra/lucene/index/PartitionedIndex.scala [hunk 27]
```diff
@@ -178,7 +178,7 @@ class PartitionedIndex(
     * @param query     the query identifying the documents to be deleted
     */
   def delete(partition: Int, query: Query) {
-    logger.debug(s"Deleting $query from $name")
+    logger.debug(s"Deleting $query from $name in partition $partition")
     indexes(partition).delete(query)
   }
```
Comment (ealonsodb) [inline plugin/src/main/scala/com/stratio/cassandra/lucene/index/PartitionedIndex.scala]: in search chunk -> partition

File: doc/documentation.rst [hunk 28]
```diff
@@ -709,6 +710,49 @@ cardinalities and uniform distributions will provide better load balancing betwe
...
+
+A virtual node based partitioner. Rows will be stored in an index partition determined by the hash of the virtual node
+token range number. Partition-directed and specific virtual node token range searches will be routed to a single partition,
+increasing performance. However, unbounded token range searches will be routed to all the partitions, with a slightly lower
+performance.
+
+Load balancing depends on virtual node token ranges distribution. The more virtual nodes, the better distribution (more
+similarity in number of tokens that falls inside any virtual node) between virtual nodes, the better load balancing.
+
+Consecutive token ranges are grouped in the same partition. This increase performance when using spark due to that spark
```
Comment (ealonsodb) [inline doc/documentation.rst]: increaseS, Apache Spark, remove this parenthesis!!!!

**Comment by adelapena** [comment]
IT should avoid concurrency issues creating indexes with different names and null column.

**Comment by adelapena** [comment]
IT should avoid concurrency issues creating indexes with different names and null column.

**Pull request merged**